### PR TITLE
Added ability to place toastview top, center or bottom, plus possibility for adjustments

### PR DESCRIPTION
**Pull request by nharbo**

### CODE REVIEW

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift [hunk 1]
```diff
@@ -118,15 +134,27 @@ extension ReachabilityConfiguration {
...
     public enum Animation {
         /// the banner will appear/dissapear from the screen by changing the view's alpha value
         case fadeInOut
-        /// the banner will appear/dissapear from the screen by moving it's frame on and off the view
-        case slideInOut
-        /// the banner will appear/dissapear from the screen by moving it's frame on and off the view while simulatiously changing it's alpha values
-        case slideAndFadeInOut
+        /// the banner will appear/dissapear from the screen by moving it's frame on and off the view from the top
+        case slideInOutFromTop
+        /// the banner will appear/dissapear from the screen by moving it's frame on and off the view from the top, while simulatiously changing it's alpha values
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift]: typo: dissapear should be disappear

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift [hunk 2]
```diff
@@ -118,15 +134,27 @@ extension ReachabilityConfiguration {
...
+        case slideAndFadeInOutFromTop
+        /// the banner will appear/dissapear from the screen by moving it's frame on and off the view from the bottom
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift]: same

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift [hunk 3]
```diff
@@ -118,15 +134,27 @@ extension ReachabilityConfiguration {
...
+        case slideInOutFromBottom
+        /// the banner will appear/dissapear from the screen by moving it's frame on and off the view from the bottom, while simulatiously changing it's alpha values
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift]: believe it or not...same 😄

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityViewController.swift [hunk 4]
```diff
@@ -65,7 +71,17 @@ class ReachabilityViewController: UIViewController {
         guard window != nil else { return }
         DispatchQueue.main.async {
             let height = self.configuration.options[.height] as! CGFloat
-            self.view.frame = CGRect(x: 0, y: -height, width: self.window.frame.width, height: height)
+            let animationType = self.configuration.options[.animation] as! ReachabilityConfiguration.Animation
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityViewController.swift]: not a big fan of `as!`. Could it be "soft" unwrapped or done differently?

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift [hunk 5]
```diff
@@ -23,51 +23,138 @@ extension Reachability.ReachabilityListener {
...
     let title: String
-    let titleColor: UIColor
-    let backgroundColor: UIColor
-    let height: CGFloat
-    let font: UIFont
-    let textAlignment: NSTextAlignment
-    let animation: Animation
+    let noConnectionTitle: String
+    let options: [ReachabilityConfiguration.Key : Any]
+    let defaultOptionsDict: [ReachabilityConfiguration.Key : Any] = [.titleColor : UIColor.white,
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift:30]: could you split this one. The formatting looks a bit odd

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift [hunk 6]
```diff
@@ -23,51 +23,138 @@ extension Reachability.ReachabilityListener {
...
+                finalOptionsDict[.appearanceAdjustment] = appearanceAdjustment
+            } else {
+                finalOptionsDict[.appearanceAdjustment] = defaultOptionsDict[.appearanceAdjustment]
+            }
+            
+        } else {
+            finalOptionsDict = defaultOptionsDict
+        }
+        
+        self.options = finalOptionsDict
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift:99]: there is a lot of code in that init method.

You could maybe say:

`var finalOptionsDict = defaultOptionsDict`

and then only replace if there was a value in `options`:

```
if let titleColor = options[.titleColor] as? UIColor {
    finalOptionsDict[.titleColor] = titleColor
} 
```

Or maybe do something with `map`?

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift [hunk 7]
```diff
@@ -23,51 +23,138 @@ extension Reachability.ReachabilityListener {
...
     }
 }
 
 extension ReachabilityConfiguration {
     
+    public enum Key: String {
+        case titleColor //Has to be of type UIColor
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityModels.swift]: Change from `//` to `///` on these comments. That way the comment will be part of the generated documentation

File: ReachabilityUI/ReachabilityUI/Reachability/ReachabilityViewController.swift [hunk 8]
```diff
@@ -82,34 +99,48 @@ class ReachabilityViewController: UIViewController {
...
             titleLabel.text = configuration.title
-            titleLabel.textColor = configuration.titleColor
-            view.backgroundColor = configuration.backgroundColor
+            titleLabel.textColor = configuration.options[.titleColor] as? UIColor
+            view.backgroundColor = configuration.options[.backgroundColor] as? UIColor
         }
     }
     
     private func animatePositionChange(_ state: State, animation: ReachabilityConfiguration.Animation) {
+        let height = self.configuration.options[.height] as! CGFloat
```
Comment (pbodsk) [inline ReachabilityUI/ReachabilityUI/Reachability/ReachabilityViewController.swift]: this func clocks in at around 70 lines! That is a lot for a function 😄 

Could parts of it be extracted to private helper methods?

**Comment by nharbo** [comment]
I made the changes @pbodsk , but from my private computer. So I couldn't push, as I don't have the ssh privatekey on this computer. So I'll have to wait until I'm back at work again :/

**Pull request merged**